xds/internal/testutils/e2e/server.go

```diff
@@ -0,0 +1,135 @@
+// envoyproxy/go-control-plane/pkg/log. This is passed to the Snapshot cache.
+type serverLogger struct{}
+
+func (l serverLogger) Debugf(format string, args ...interface{}) { logger.Infof(format, args...) }
+func (l serverLogger) Infof(format string, args ...interface{})  { logger.Infof(format, args...) }
+func (l serverLogger) Warnf(format string, args ...interface{})  { logger.Warningf(format, args...) }
+func (l serverLogger) Errorf(format string, args ...interface{}) { logger.Errorf(format, args...) }
+
+// ManagementServer is a thin wrapper around the xDS control plane
+// implementation provided by envoyproxy/go-control-plane.
+type ManagementServer struct {
+	// Address is the host:port on which the management server is listening for
+	// new connections.
+	Address string
+
+	version int
+	gs      *grpc.Server
+	xs      v3server.Server
+	cache   v3cache.SnapshotCache
+}
+
+// StartManagementServer initializes a management server which implements the
+// AggregatedDiscoveryService endpoint. The management server is initialized
+// with no resources. Tests should call the Update() method to change the
+// resource snapshot held by the management server, as required by the test
+// logic. When the test is done, it should call the Stop() method to cleanup
+// resources allocated by the management server.
+func StartManagementServer(ctx context.Context) (*ManagementServer, error) {
+	// Create a snapshot cache.
+	cache := v3cache.NewSnapshotCache(true, v3cache.IDHash{}, serverLogger{})
+	logger.Infof("Created new snapshot cache...")
+
+	// Create an xDS management server and register the ADS implementation provided by it rough a gRPC server.
+	xs := v3server.NewServer(ctx, cache, v3server.CallbackFuncs{})
+	gs := grpc.NewServer()
+	v3discoverygrpc.RegisterAggregatedDiscoveryServiceServer(gs, xs)
+	logger.Infof("Registered Aggregated Discovery Service (ADS)...")
+
+	// Start serving.
+	lis, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return nil, fmt.Errorf("failed to start xDS management server: %v", err)
+	}
+	go gs.Serve(lis)
+	logger.Infof("xDS management server serving at: %v...", lis.Addr().String())
+
+	return &ManagementServer{
+		Address: lis.Addr().String(),
+		version: 0,
+		gs:      gs,
+		xs:      xs,
+		cache:   cache,
+	}, nil
+}
+
+// UpdateOptions wraps parameters to be passed to the Update() method.
+type UpdateOptions struct {
+	// NodeID is the id of the client to which this update is to be pushed.
+	NodeID string
+	// Listeners is the updated list of listener resources.
+	Listeners []*v3listenerpb.Listener
+	// TODO(easwars): Add support for other resource types.
+}
+
+// Update changes the resource snapshot held by the management server, which
+// updates connected clients as required.
+func (s *ManagementServer) Update(opts UpdateOptions) error {
+	s.version++
+
+	// Create a snapshot with the passed in resources.
+	var listeners []types.Resource
+	if opts.Listeners != nil {
```

Comment:
No need for this `if`.
`for` works on `nil` slice.

Reply:
facepalm